7057 export import visualization data

juanignaciosl opened the pull request:
This closes #7057.

Some context:
- Feature request: #6365
- This allows exporting a full visualization (metadata + data) with a rake to a `.carto` (zipped) file.
- This enabling importing it with drag & drop.
- Next steps:
  - Export through API #7125.
  - Geopackage format #7262.

Cartofante commented:
Can one of the admins verify this patch?

juanignaciosl commented:
@rafatower could you CR this, please? I'll request a second CR to a Editor team backender, but since this involves the importer I'd like you to take the first look. Ignore the "Carto::DataExporter#export_tables exports a table in requested format" test for the moment.

juanignaciosl commented:
Retest this.

juanignaciosl commented:
Retest this, please.

PS: only Travis build failed, because of `npm install` indeed.

juanignaciosl commented:
Retest this, please.

rafatower commented:
@iriberri I think it is good to have your :eyes: here too

rafatower commented:
:+1:

iriberri commented:
Looks neat! :-)

juanignaciosl commented:
Thank you both, @rafatower and @iriberri!

Could you please CR this, @javitonino ?

@luisbosque Visualization export will use a tmp directory, `/tmp/exporter` by default. Is there something extra that should be done to ensure that it gets cleaned? Currently this is only a rake that must be run locally to generate the file. #7125 will move this to a queue and will also probably use S3 for storage.

luisbosque commented:
Ey @juanignaciosl there should not be any additional step. We already have a cron task in editor servers than keep cleaning /tmp

juanignaciosl commented:
@luisbosque oído cocina. Thanks!

javitonino commented on app/connectors/importer.rb:
```diff
@@ -99,16 +97,32 @@ def create_overviews(result)
+          user = Carto::User.find(data_import.user_id)
+          runner.visualizations.each do |visualization|
+            vis = Carto::VisualizationsExportPersistenceService.new.save_import(user, visualization)
+            bind_visualization_to_data_import(vis)
```
Only the last visualization will be bound. Is it possible to import more than one visualization per run?

juanignaciosl replied:
Currently it's not possible, only one visualization is exported. Something similar happens with tables, `DataImport` was meant for one table and we're pushing it too far without a redesign. Anyway since DataImport is only relevant to trace the state of the import process there's no too much value on the DataImport --> Visualization relationship.

javitonino replied:
:sparkles:

javitonino commented on app/models/data_import.rb:
```diff
@@ -360,6 +360,16 @@ def table
 
+  def tables
+    table_names_array.map do |table_name|
+      ::Table.new(user_table: UserTable.where(name: table_name, user_id: user_id).first)
```
I think using the `service` method in `UserTable` is easier to read:

`UserTable.where(name: table_name, user_id: user_id).first.service`

juanignaciosl replied:
Agree 👌

javitonino commented on config/app_config.yml.sample:
```diff
@@ -170,6 +170,8 @@ defaults: &defaults
       api_key: ''
       table_name: ''
+  exporter:
+    exporter_temporal_folder: '/tmp/exporter'
```
Although unneeded because this is the same as the default, maybe you can add this to `app_config.yml.testing` (used by the testing scripts).

juanignaciosl replied:
👌

javitonino commented on app/models/carto/visualization_export.rb:
```diff
@@ -0,0 +1,101 @@
+    end
+
+    def ensure_folder(folder)
+      Dir.mkdir(folder) unless Dir.exists?(folder)
```
This only creates the last folder, it is non-recursive. It will fail if the parent directory does not exists. You can use [`FileUtils.mkdir_p`](http://ruby-doc.org/stdlib-2.2.3/libdoc/fileutils/rdoc/FileUtils.html#method-c-mkdir_p) instead.

juanignaciosl replied:
true 👌

javitonino commented on app/models/carto/visualization_export.rb:
```diff
@@ -0,0 +1,101 @@
+    end
+
+    def export_visualization_tables(visualization, dir, format)
+      visualization.related_tables.map { |ut| export_table(ut, dir, format) }
```
This exports all tables used in the visualization, but you only export layers that the user has access to: `layers_with_data_readable_by(user)`. Is this intentional? Shouldn't this make the same check for readability?

juanignaciosl replied:
Good catch! Fix in progress...

javitonino commented on app/models/carto/layer.rb:
```diff
@@ -43,6 +43,10 @@ def affected_tables
     end
 
+    def data_readable_by?(user)
+      affected_tables.find { |ut| !ut.readable_by?(user) }.nil?
```
Maybe `affected_tables.all? { |ut| ut.readable_by?(user) }`?

juanignaciosl replied:
The nice thing about `find` is that it's more efficient, since it will stop on the first occurrence, at the cost of being slightly more uncomfortable to read. It can be changed it if you prefer though.

javitonino replied:
I have just tested `all?` and it also stops on the first `false` or `nil` (although I cannot find anything in the documentation). Anyway, up to you.

```
tt = Proc.new { puts 'true'; true }
ff = Proc.new { puts 'false'; false }
[tt, tt, ff, ff, tt].all?{ |p| p.call }
true
true
false
=> false
```

juanignaciosl replied:
Oh, changing then :-) My point was based on a comment, not proofs :_) Thanks!

javitonino commented on app/models/carto/visualization_export.rb:
```diff
@@ -0,0 +1,101 @@
+      ensure_folder(exporter_config[:exporter_temporal_folder] || DEFAULT_EXPORTER_TMP_FOLDER)
+    end
+
+    def export_dir(visualization, base_dir: DEFAULT_EXPORTER_TMP_FOLDER)
```
Why not `base_dir: exporter_folder`?

juanignaciosl replied:
Oh, true 👍

javitonino commented on app/models/carto/visualization_export.rb:
```diff
@@ -0,0 +1,101 @@
+      File.open(visualization_json_file, 'w') { |file| file.write(visualization_json) }
+
+      zipfile = "#{visualization_id}.carto"
+      `cd #{export_dir}/ && zip -r #{zipfile} #{visualization_id} && cd -`
```
I don't know if following the installation instructions installs `zip` or if it is part of a default Ubuntu 12.04 installation. It is not the case with Arch Linux, but that's expected.

We should consider documenting the `zip` dependency somewhere if it is not already there. Alternatively we could use a gem like [rubyzip](https://github.com/rubyzip/rubyzip) to avoid this altogether.

juanignaciosl replied:
I first considered using `rubyzip`, but I rejected it:
- It's too complicated (please check the instructions for ("Zipping a directory recursively")[https://github.com/rubyzip/rubyzip#zipping-a-directory-recursively]).
- It works in memory (at least it did, I saw some absurd benchmarks about it).

I also considered `zipruby` but after a couple of test using a command line seemed to be a much natural approach, since CartoDB uses other commands.

@luisbosque is this ok for you? What's the right approach for documenting this dependency? Updating NEWS.md + installation instructions?

javitonino replied:
:+1:, I have no problem with using commands in the importer as it already uses them and it's easier.

javitonino commented on app/models/carto/visualization_export.rb:
```diff
@@ -0,0 +1,101 @@
+               base_dir: exporter_folder)
+      visualization_id = visualization.id
+      export_dir = export_dir(visualization, base_dir: base_dir)
+      tmp_dir = tmp_dir(visualization, base_dir: base_dir)
```
Maybe call `ensure_clean_folder` on `tmp_dir`. If there are remains from a previous execution, they will be added to the zip.

juanignaciosl replied:
Ok. I'm also making some part of the route random, because it's technically possible that a visualization gets exported more than once, even by the same user.

juanignaciosl replied:
PD: and remove the use of `export_dir` method _inside_ `tmp_dir` method, using it as default parameter instead. More readable, that hidden dependency made no sense.

javitonino commented on spec/connectors/importer_spec.rb:
```diff
@@ -313,4 +314,59 @@
+        append: false,
+        create_visualization: true
+      )
+      data_import.values[:data_source] = filepath
```
¯\_(ツ)_/¯

javitonino commented on spec/connectors/importer_spec.rb:
```diff
@@ -313,4 +314,59 @@
+      # Fixture file checks
+      data_import.table_names.should eq "guess_country twitter_t3chfest_reduced"
+      visualization = Carto::Visualization.find(data_import.visualization_id)
+      visualization.name.should eq "map with two layers"
```
Can you check here that both layers were imported?

juanignaciosl replied:
Sure.

javitonino commented on spec/factories/users_helper.rb:
```diff
@@ -23,3 +23,21 @@
+shared_context 'user helper' do
+  before(:all) do
+    username = "a#{String.random(10)}-a".downcase
+    @user = create_user(email: "#{username}@cartotest.com", username: username, password: '123456')
```
Why not use `UniqueNamesHelper` here? Also, any reason for using `create_user` over the `:valid_user` factory?

juanignaciosl replied:
Changed to `:valid_user`. I've just pushed all the changes, maybe you want to take a second look.

javitonino commented:
Looks good after the changes. Great work on this. :package: it!

javitonino commented:
Forgot to mention, it would be nice to add `/tmp/exporter` to the servers configuration, although not needed as there is already a default. If you decide to do it, ping me for help if you need, as I just did it yesterday for another PR.

juanignaciosl commented:
@iriberri , @javitonino , would you do a second CR, please? I've fixed the infamous issue with URLs including `.carto` and also fixed #7279, which I overlooked previously.

javitonino commented on app/services/carto/visualizations_export_persistence_service.rb:
```diff
@@ -64,5 +64,12 @@ def save_import(user, visualization)
+
+    def apply_user_limits(user, visualization)
+      visualization.privacy = Carto::Visualization::PRIVACY_PUBLIC unless user.private_maps_enabled
+      visualization.map.layers = visualization.map.layers.take(user.max_layers)
```
Basemap layers don't count towards the limit, only data layers (`carto_and_torque_layers`) do. So I'm afraid you need a little more logic here.

juanignaciosl replied:
Changed, take a look if you want. Good catch!

javitonino commented:
Changes look good to me. :+1: after fixing the layer limit check.

javitonino commented:
:sparkles:

iriberri commented:
LGTM :+1:

juanignaciosl commented on services/importer/lib/importer/downloader.rb:
```diff
@@ -98,10 +98,15 @@ def self.supported_extensions
                                   .sort_by(&:length).reverse
       end
 
+      def self.supported_extensions_match
```
@xavijam could you CR the frontend side, please? I'm _almost_ sure that this change comes from a wrong merge, since I only wanted to add the previous extension.

xavijam commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cartodb-ui",
-  "version": "3.23.70",
```
Do you need this?

xavijam replied:
Yep, I've just seen we need to update the upload-config :+1:

xavijam commented on lib/assets/javascripts/cartodb/common/background_polling/models/upload_config.js:
```diff
@@ -37,7 +37,8 @@ module.exports = {
     'sql',
-    'rar'
+    'rar',
+    'carto'
```
👍
